crates/core/tedge_api/src/device_id.rs

```diff
@@ -0,0 +1,36 @@
+pub fn new_get_child_id_from_topic(parent_device_name: String, topic: String) -> Option<String> {
+    if topic.starts_with("te/device") {
+        let topic_split = topic.split('/').collect::<Vec<_>>();
+        if topic_split.len() >= 5 {
+            if topic_split[2].eq("main") {
+                return None;
+            }
+
+            let mut child_device_name = parent_device_name;
+            for item in topic_split.iter().take(5) {
+                if item.ne(&"te") && !item.is_empty() {
+                    child_device_name.push_str(&format!(":{}", item));
+                }
+            }
+            Some(child_device_name)
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+    #[test_case("test_device".to_string(), "te/device/main///m/".to_string(), None; "main device id")]
+    #[test_case("test_device".to_string(), "te/device/child///m/".to_string(), Some("test_device:device:child".to_string()); "child device id")]
+    #[test_case("test_device".to_string(), "te/device/child/a1/a2/m/".to_string(), Some("test_device:device:child:a1:a2".to_string()); "child with more components device id")]
+    #[test_case("test_device".to_string(), "foo/bar".to_string(), None; "wrong topic")]
```

Comment:
Let's define what the expected cases and  error cases for the target function as of now.
It needs to generate the c8y acceptable unique device name from the given topic name. However, this logic will be changed once we have the entity store. The mapper will just need to query.

My suggestion is to accept only `<root>/device/<something>/service/<something>` for the first 5 digits.

So coverage will be:
* te/device/main///m/ => test_device
* te/device/child///m/ => test_device:device:child
* te/device/main/service/nodered/m/ => test_device:device:main:service:nodered
* te/device//service/nodered/m => error
* te/device/main/service//m => error

Reply:
IMO the `telemetry` for the service can be addressed as part of another PR. As of now, thin-edge does not support the telemetry for services, isn't it?